numpy/core/src/multiarray/textreading/rows.c

```diff
@@ -431,6 +492,12 @@ read_rows(stream *s,
         data_ptr += row_size;
     }
 
+    if (usecols_obj != Py_None) {
+        // This function owns usecols if a callable usecols_obj was given.
+        PyMem_Free(usecols);
+        usecols = NULL;  // An overabundance of caution...
+    }
```

Comment:
Would it be ugly to move all usecols handling here?  Its fine, but conditional ownership tens to be an anti-pattern.

Another way to spell would be to have `usecol_from_func = ...` and `usecol = usecol_from_func`, so the owner is unconditionally `usecol_from_func` and not "usecol" itself.

Reply:
> Would it be ugly to move all usecols handling here? 

No, I think that would be fine.  In fact, with a bit more of the argument processing code moved into `read_rows()`, I think the function `_readtext_from_stream()` can be eliminated, and `_load_from_filelike()` can call `read_rows()` directly.  I'll give that shot, and push to the PR if it looks reasonable.